File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.util.*;
class Solution {
    public String[] solution(String[] players, String[] callings) {
        String[] answer = new String[players.length];
        Map<String, Integer> mp1 = new HashMap<>(); // 이름 - 등수
        Map<Integer, String> mp2 = new HashMap<>(); // 등수 - 이름
        for(int i = 0; i < players.length; i++) {
                mp1.put(players[i], i + 1);
                mp2.put(i + 1, players[i]);
        };
        
        for(String s : callings){
            int idx = mp1.get(s);
            String temp = mp2.get(idx - 1);
            mp1.put(s, idx - 1); // 추월한놈 빼주기
            mp1.put(temp, idx); // 당한놈 그자리 대체
            mp2.put(idx - 1, s); // 추월한 놈 등수 빼주기
            mp2.put(idx, temp);
        }
        
        for(int i = 0; i < answer.length; i++) answer[i] = mp2.get(i + 1);
    
        return answer;
    }
}
/*
1등~ 현재등수 players
부른 이름 callings

*/

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    rank = {x:i for i,x in enumerate(players)}        
    for call in callings:
        now = rank[call] # 추월한놈 현재등수
        temp = players[now-1] # 추월 당할 놈 이름
        players[now-1] = players[now]
        players[now] = temp
        rank[call] = now-1
        rank[temp] = now
    return players
    
    
"""
[조건]
바로 앞 추월할 때 추월한 선수 이름 부름 
이름을 부르면 바로 앞에 놈이랑 자리를 바꾼다

[아이디어]
이름-등수 맵에 저장 이름 부를 때마다 나보다 하나 앞에 있는 놈이랑 스왑하면서 이름 - 등수 바꾸기


"""